repa/grids/psdiffusion.cpp

```diff
@@ -0,0 +1,158 @@
+/**
+ * Copyright 2017-2019 The repa authors
+ *
+ * This file is part of Repa.
+ *
+ * Repa is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Repa is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Repa.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "psdiffusion.hpp"
+
+#include <algorithm>
+#include <set>
+
+#include "grids/util/mpi_cart.hpp"
+#include "grids/util/mpi_graph.hpp"
+#include "util/vec_arith.hpp"
+
+#ifndef NDEBUG
+#define PSDIFFUSION_DEBUG
+#endif
+
+namespace repa {
+namespace grids {
+
+static Vec3i
+fix_periodic_edge(const Vec3i &c0, const Vec3i &c2, const Vec3i &gsize)
+{
+    using namespace util::vector_arithmetic;
+    Vec3i te, ts;
+    for (int i = 0; i < te.size(); i++) {
+        te[i] = c2[i] - c0[i] > 1;
+        ts[i] = c0[i] - c2[i] > 1;
+    }
+    return c2 + ts * gsize - te * gsize;
+}
+
+/*
+ * Initialization
+ */
+PSDiffusion::PSDiffusion(const boost::mpi::communicator &comm,
+                         Vec3d box_size,
+                         double min_cell_size)
+    : Diffusion(comm, box_size, min_cell_size)
+{
+}
+
+PSDiffusion::~PSDiffusion()
+{
+}
+
+void PSDiffusion::post_init(bool firstcall)
+{
+    Diffusion::post_init(firstcall);
+    comm_dims = util::mpi_cart_get_dims(comm_cart);
+
+#ifdef PSDIFFUSION_DEBUG
```

Comment:
Please, quickly comment here, that we save the neighborhood to check it later, don't we?

Reply:
Updated in 14f7a25